server/parsec/components/memory/shamir.py

```diff
@@ -0,0 +1,47 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+from parsec._parsec import DeviceID, OrganizationID, UserID, VerifyKey, authenticated_cmds
+from parsec.components.events import EventBus
+from parsec.components.memory.datamodel import (
+    MemoryDatamodel,
+    MemoryShamirSetup,
+)
+from parsec.components.shamir import (
+    BaseShamirComponent,
+    VerifyCertificatesBadOutcome,
+    verify_certificates,
+)
+
+
+class MemoryShamirComponent(BaseShamirComponent):
+    def __init__(self, data: MemoryDatamodel, event_bus: EventBus) -> None:
+        super().__init__()
+        self._data = data
+        self._event_bus = event_bus
+
+    async def remove_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+    ) -> None:
+        self._data.organizations[organization_id].shamir_setup.pop(author)
+
+    async def add_recovery_setup(
+        self,
+        organization_id: OrganizationID,
+        author: UserID,
+        device: DeviceID,
+        author_verify_key: VerifyKey,
+        setup: authenticated_cmds.latest.shamir_recovery_setup.ShamirRecoverySetup,
+    ) -> None | authenticated_cmds.latest.shamir_recovery_setup.Rep:
+        match verify_certificates(setup, device, author_verify_key):
+            case (brief, shares):
+                if self._data.organizations[organization_id].shamir_setup.get(author) is None:
+                    self._data.organizations[organization_id].shamir_setup[author] = (
+                        MemoryShamirSetup(setup.ciphered_data, setup.reveal_token, brief, shares)
+                    )
+                else:
+                    return authenticated_cmds.latest.shamir_recovery_setup.RepAlreadySet()
+
+            case VerifyCertificatesBadOutcome.INVALID_DATA:
+                return authenticated_cmds.latest.shamir_recovery_setup.RepInvalidData()
```

Comment:
Contrary to v2, this method needs to include all the required checks:
- organization exists and not expired
- author exists and not revoked and not frozen
- all recipients exists and not revoked
- check that certificate timestamps are strictly increasing in the shamir topic

See methods in other components for reference.